### PR TITLE
[zh-cn]: Update overview of Classes

### PR DESCRIPTION
**Pull request by BSPR0002**
### Related issues and pull requests
Relates to #18329

### CODE REVIEW

File: files/zh-cn/web/javascript/reference/classes/index.md [hunk 1]
```diff
@@ -5,106 +5,135 @@ slug: Web/JavaScript/Reference/Classes
 
 {{JsSidebar("Classes")}}
 
-类是用于创建对象的模板。他们用代码封装数据以处理该数据。JS 中的类建立在原型上，但也具有某些语法和语义未与 ES5 类相似语义共享。
+类是用于创建对象的模板。它们用代码封装数据用以处理数据。JS 中的类建立在[原型](/zh-CN/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)之上，同时还具有一些类独有的语法和语义。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/classes/index.md:8]:
```suggestion
类是用于创建对象的模板。它们用代码封装数据以对其进行处理。JS 中的类建立在[原型](/zh-CN/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)之上，同时还具有一些类独有的语法和语义。
```

File: files/zh-cn/web/javascript/reference/classes/index.md [hunk 2]
```diff
@@ -5,106 +5,135 @@ slug: Web/JavaScript/Reference/Classes
...
     this.height = height;
     this.width = width;
   }
 };
-console.log(Rectangle.name);
-// 输出："Rectangle2"
 ```
 
-> **备注：** 类**表达式**也同样受到[类声明](/zh-CN/docs/Web/JavaScript/Reference/Classes#Class_declarations)部分中提到的类型提升的限制。
+类似于函数表达式，类表达式可以是匿名的，或者也可以有一个不同于被赋值给的变量的名称的名字。然而，不同于函数声明的是，类声明具有与 `let` 和 `const` 相同的[暂时性死区](/zh-CN/docs/Web/JavaScript/Reference/Statements/let#temporal_dead_zone_tdz)限制，并且表现得像是[没有被提升](/zh-CN/docs/Web/JavaScript/Guide/Using_classes#类声明提升)一样。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/classes/index.md:44]:
```suggestion
似于函数表达式，类表达式可以是匿名的，或者也可以有一个不同于被赋值给的变量的名称的名字。然而，不同于函数声明的是，类声明具有与 `let` 和 `const` 相同的[暂时性死区](/zh-CN/docs/Web/JavaScript/Reference/Statements/let#temporal_dead_zone_tdz)限制，并且表现得像是[没有被提升](/zh-CN/docs/Web/JavaScript/Guide/Using_classes#类声明提升)一样。
```

File: files/zh-cn/web/javascript/reference/classes/index.md [hunk 3]
```diff
@@ -5,106 +5,135 @@ slug: Web/JavaScript/Reference/Classes
...
 
-### 原型方法
+> **备注：** 私有属性具有在同一个类种声明的所有属性的名称必须唯一的限制。其他所有的公有属性都没有这个限制——你可以写多个同名的公有属性，但是最后一个会覆盖掉其他的。此种表现与[对象初始化器](/zh-CN/docs/Web/JavaScript/Reference/Operators/Object_initializer#重复属性名)相同。
 
-参见[方法定义](/zh-CN/docs/Web/JavaScript/Reference/Functions/Method_definitions)。
+另外，有两种特殊的类元素语法：[`constructor`](#构造函数) 和[静态初始化块](#静态初始化块)，它们有自己的参考资料。
+
+#### 构造函数
+
+[constructor](/zh-CN/docs/Web/JavaScript/Reference/Classes/constructor) 方法是用于创建和初始化一个由类创建的对象的特殊方法。一个类只能拥有一个名为“constructor”的特殊方法。如果类包含多个 `constructor` 方法，将会抛出一个 {{jsxref("SyntaxError")}}。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/classes/index.md:79]:
```suggestion
{{jsxref("Classes/constructor", "constructor")}} 方法是用于创建和初始化一个由类创建的对象的特殊方法。一个类只能拥有一个名为“constructor”的特殊方法。如果类包含多个 `constructor` 方法，将会抛出一个 {{jsxref("SyntaxError")}}。
```

File: files/zh-cn/web/javascript/reference/classes/index.md [hunk 4]
```diff
@@ -114,103 +143,30 @@ class Point {
...
-
-Animal.eat(); // class Animal
-let eat = Animal.eat;
-eat(); // undefined
+console.log(Point.displayName); // "Point"
+console.log(Point.distance(p1, p2)); // 7.0710678118654755
 ```
 
-如果上述代码通过传统的基于函数的语法来实现，那么依据初始的 _this_ 值，在非严格模式下方法调用会发生自动装箱。若初始值是 `undefined`，_this_ 值会被设为全局对象。
+#### 字段的声明
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/classes/index.md:165]:
```suggestion
#### 字段声明
```

File: files/zh-cn/web/javascript/reference/classes/index.md [hunk 5]
```diff
@@ -221,17 +177,17 @@ class Rectangle {
...
 ```
 
-通过预先声明字段，类定义变得更加自我记录，并且字段始终存在。
+类字段与对象属性相似，不属于变量，所以我们不需要使用诸如 `const` 一类的关键字去声明它们。在 JavaScript 中，[私有属性](#私有属性_2)使用了一种特殊的语法，所以不应当使用像 `public` 和 `private` 这样的修饰符关键字。
 
-在声明字段时，我们不需要像 `let`、`const` 和 `var` 这样的关键字。
+如上文所示，声明字段时有没有默认值都可以。没有默认值的字段其值默认为 `undefined`。通过预先声明字段，类声明会变得自我文档化，并且字段始终显现，有助于优化代码。
 
-正如上面看到的，这个字段可以用也可以不用默认值来声明。
+参见[类公有字段](/zh-CN/docs/Web/JavaScript/Reference/Classes/Public_class_fields)获取更多信息。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/classes/index.md:184]:
```suggestion
参见[公有类字段](/zh-CN/docs/Web/JavaScript/Reference/Classes/Public_class_fields)获取更多信息。
```

File: files/zh-cn/web/javascript/reference/classes/index.md [hunk 6]
```diff
@@ -242,17 +198,15 @@ class Rectangle {
...
+私有字段只能在字段声明中预先声明。它们不像普通属性那样可以通过赋值创建。
 
-更多信息，请看[class fields](/zh-CN/docs/Web/JavaScript/Reference/Classes/Class_fields).
+更多信息，参见[私有属性](/zh-CN/docs/Web/JavaScript/Reference/Classes/Private_properties)。
 
-## 使用 `extends` 扩展子类
+### 继承
 
-[`extends`](/zh-CN/docs/Web/JavaScript/Reference/Classes/extends) 关键字在 _类声明_ 或 _类表达式_ 中用于创建一个类作为另一个类的一个子类。
+{{jsxref("Classes/extends", "extends")}} 关键字用于*类声明*或*类表达式*中，用以创建一个作为另一构造函数（类或函数）的子类的类。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/classes/index.md:209]:
```suggestion
{{jsxref("Classes/extends", "extends")}} 关键字用于*类声明*或*类表达式*中，用以创建一个作为另一构造函数（类或函数）的子类。
```

File: files/zh-cn/web/javascript/reference/classes/index.md [hunk 7]
```diff
@@ -261,136 +215,112 @@ class Animal {
...
+当一个 [`class` 声明](/zh-CN/docs/Web/JavaScript/Reference/Statements/class)或 [`class` 表达式](/zh-CN/docs/Web/JavaScript/Reference/Operators/class)被求值，它的各个组件将按照以下顺序被求值：
 
-例如，当使用像{{jsxref("Array.map", "map()")}}返回默认构造函数的方法时，你希望这些方法返回一个父`Array`对象，而不是`MyArray`对象。{{jsxref("Symbol.species")}} 符号可以让你这样做：
+1. {{jsxref("Classes/extends", "extends")}} 子句，如果有，将会首先被求值。它必须被求取为一个合法的构造函数或 `null`，否则将抛出一个 {{jsxref("TypeError")}}。
+2. 提取 {{jsxref("Classes/constructor", "constructor")}} 方法，如果 `constructor` 不存在将会用默认实现进行替换。但是，因为 `constructor` 的定义仅仅只是一个方法的定义，所以这一步是观察不到的。
+3. 按照声明顺序对类元素的属性键名求值。如果属性键名是计算属性名，则对表达式求值，表达式中的 `this` 指向类声明所处上下文的 `this`（不是类本身），属性值尚不会被求值。
+4. 按照声明顺序安设方法和访问器。实例方法和访问器被安设在当前类的 `prototype` 属性上，静态方法和访问器被安设在类本身。私有实例方法和访问器会被保存，之后会直接安置到实例上，这个步骤不可被观察到。
+5. 类现在已经用 `extends` 指定的原型和 `constructor` 指定的实现初始化完成。对于上面的所有步骤，如果有表达式尝试访问类名，会抛出一个 {{jsxref("ReferenceError")}}，因为类还没有初始化完成。
+6. 按照声明顺序求取类元素的值：
+   - 对与每个[实例字段](/zh-CN/docs/Web/JavaScript/Reference/Classes/Public_class_fields) (public or private)，其初始化器表达式会被保存。初始化器会在实例创建期间被求值，时间点在构造函数开头（对于基类）或者在调用 `super()` 返回时立刻求值（对于派生类）。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/classes/index.md:272]:
```suggestion
   - 对与每个[实例字段](/zh-CN/docs/Web/JavaScript/Reference/Classes/Public_class_fields)（共有或私有），其初始化器表达式会被保存。初始化器会在实例创建期间被求值，时间点在构造函数开头（对于基类）或者在调用 `super()` 返回时立刻求值（对于派生类）。
```

File: files/zh-cn/web/javascript/reference/classes/index.md [hunk 8]
```diff
@@ -401,19 +331,10 @@ class Bar extends calculatorMixin(randomizerMixin(Foo)) {}
...
-- {{jsxref("Operators/class", "类表达式", "", "true")}}
-- {{jsxref("Classes/Public_class_fields", "公有类字段", "", "true")}}
-- {{jsxref("Classes/Private_properties", "私有属性", "", "true")}}
-- {{jsxref("Operators/super", "super")}}
-- [Blog post: "ES6 In Depth: Classes"](https://hacks.mozilla.org/2015/07/es6-in-depth-classes/)
-- [Fields and public/private class properties proposal (stage 3)](https://github.com/tc39/proposal-class-fields)
+- [使用类](/zh-CN/docs/Web/JavaScript/Guide/Using_classes)教程
+- [`class`](/zh-CN/docs/Web/JavaScript/Reference/Statements/class)
+- [`class` 表达式](/zh-CN/docs/Web/JavaScript/Reference/Operators/class)
+- [Functions](/zh-CN/docs/Web/JavaScript/Reference/Functions)
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/classes/index.md:339]:
```suggestion
- [函数](/zh-CN/docs/Web/JavaScript/Reference/Functions)
```

**Review by yin1999 — COMMENTED** [review]
Almost looks good to me. Just some nits:

File: files/zh-cn/web/javascript/reference/classes/index.md (hunk 2 above)
Comment (BSPR0002) [reply files/zh-cn/web/javascript/reference/classes/index.md:44]: I prefer this
```suggestion
与函数表达式类似，类表达式可以是匿名的，或者也可以有一个不同于被赋值给的变量的名称的名字。然而，不同于函数声明的是，类声明具有与 `let` 和 `const` 相同的[暂时性死区](/zh-CN/docs/Web/JavaScript/Reference/Statements/let#temporal_dead_zone_tdz)限制，并且表现得像是[没有被提升](/zh-CN/docs/Web/JavaScript/Guide/Using_classes#类声明提升)一样。
```

File: files/zh-cn/web/javascript/reference/classes/index.md (hunk 7 above)
Comment (BSPR0002) [reply files/zh-cn/web/javascript/reference/classes/index.md:272]:
```suggestion
   - 对于每个[实例字段](/zh-CN/docs/Web/JavaScript/Reference/Classes/Public_class_fields)（共有或私有），其初始化器表达式会被保存。初始化器会在实例创建期间被求值，时间点在构造函数开头（对于基类）或者在调用 `super()` 返回时立刻求值（对于派生类）。
```

File: files/zh-cn/web/javascript/reference/classes/index.md (hunk 6 above)
Comment (BSPR0002) [reply files/zh-cn/web/javascript/reference/classes/index.md:209]: 如果想贴近原版，那么应该是这样：
```suggestion
{{jsxref("Classes/extends", "extends")}} 关键字用于*类声明*或*类表达式*中，用以创建一个类作为另一构造函数（类或函数）的子类。
```

**Comment by yin1999** [comment]
可以

File: files/zh-cn/web/javascript/reference/classes/index.md [hunk 9]
```diff
@@ -5,106 +5,135 @@ slug: Web/JavaScript/Reference/Classes
...
-  // constructor
   constructor(height, width) {
     this.height = height;
     this.width = width;
   }
-  // Getter
+}
+```
+
+或者，如果实例属性的值不依赖构造函数的参数，那么你可以把它们定义为[类字段](#字段的声明)。
```
Comment (BSPR0002) [inline files/zh-cn/web/javascript/reference/classes/index.md:94]:
```suggestion
或者，如果实例属性的值不依赖构造函数的参数，那么你可以把它们定义为[类字段](#字段声明)。
```

**Review by yin1999 — APPROVED** [review]
:+1:

**Pull request merged**